src/workbench/Workbench.tsx

```diff
@@ -23,43 +23,47 @@ import { SizedMode } from "../common/SplitView/SplitView";
 import { ConnectionStatus } from "../device/device";
 import { useConnectionStatus } from "../device/device-hooks";
 import EditorArea from "../editor/EditorArea";
-import { MAIN_FILE } from "../fs/fs";
+import { FileVersion, MAIN_FILE } from "../fs/fs";
 import { useProject } from "../project/project-hooks";
 import ProjectActionBar from "../project/ProjectActionBar";
 import SerialArea from "../serial/SerialArea";
 import { useSettings } from "../settings/settings";
 import Simulator from "../simulator/Simulator";
 import Overlay from "./connect-dialogs/Overlay";
 import SideBar from "./SideBar";
-import { useSelection } from "./use-selection";
+import { WorkbenchSelection, useSelection } from "./use-selection";
 import { flags } from "../flags";
 
 const minimums: [number, number] = [380, 580];
 const simulatorMinimums: [number, number] = [275, 0];
 
+const defaultSelection = (
+  selection: WorkbenchSelection,
+  files: FileVersion[]
+) => {
+  // Selected file deleted? Default it.
+  if (!files.find((x) => x.name === selection.file) && files.length > 0) {
+    const defaultFile = files.find((x) => x.name === MAIN_FILE) ?? files[0];
+    return { file: defaultFile.name, location: { line: undefined } };
+  }
+  return selection;
+};
+
 /**
  * The main app layout with resizable panels.
  */
 const Workbench = () => {
-  const [selection, setSelection] = useSelection();
   const intl = useIntl();
+
+  const [maybeInvalidSelection, setSelection] = useSelection();
   const { files } = useProject();
+  const selection = defaultSelection(maybeInvalidSelection, files);
   const setSelectedFile = useCallback(
     (file: string) => {
       setSelection({ file, location: { line: undefined } });
     },
     [setSelection]
   );
-  useEffect(() => {
```

Comment:
I think this was a listener ordering thing. I've changed it to avoid making a state update at all.